src/NuGet.Jobs.GitHubIndexer/GitRepoSearchers/GitHub/GitHubSearcher.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using NuGetGallery;
+using Octokit;
+
+namespace NuGet.Jobs.GitHubIndexer
+{
+    public class GitHubSearcher : IGitRepoSearcher
+    {
+        private readonly IGitHubClient _client;
+        private readonly ILogger<GitHubSearcher> _logger;
+        private readonly TimeSpan OneMinute = TimeSpan.FromMinutes(1);
+        private readonly IOptionsSnapshot<GitHubSearcherConfiguration> _configuration;
+        private readonly IGitHubSearchApiRequester _searchApiRequester;
+
+        private DateTimeOffset _throttleResetTime;
+
+        public GitHubSearcher(
+            IGitHubClient client,
+            IGitHubSearchApiRequester searchApiRequester,
+            ILogger<GitHubSearcher> logger,
+            IOptionsSnapshot<GitHubSearcherConfiguration> configuration)
+        {
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _searchApiRequester = searchApiRequester ?? throw new ArgumentNullException(nameof(searchApiRequester));
+        }
+
+        private int _minStars => _configuration.Value.MinStars;
+        private int _resultsPerPage => _configuration.Value.ResultsPerPage;
+        private int _maxGithubResultPerQuery => _configuration.Value.MaxGitHubResultPerQuery;
+
+        private async Task CheckThrottle()
+        {
+            if (_client.GetLastApiInfo() != null && _client.GetLastApiInfo().RateLimit.Remaining == 0)
+            {
+                //var sleepTime = _client.GetLastApiInfo().RateLimit.Reset - DateTimeOffset.Now;
+                var sleepTime = _throttleResetTime - DateTimeOffset.Now;
+                _throttleResetTime = DateTimeOffset.Now;
+                _logger.LogInformation($"Waiting {sleepTime.TotalSeconds} seconds to cooldown.");
+                if (sleepTime.TotalSeconds > 0)
+                {
+                    await Task.Delay(sleepTime);
+                }
+
+                _logger.LogInformation("Resuming search.");
+            }
+        }
+
+        private async Task<SearchRepositoryResult> SearchRepo(SearchRepositoriesRequest request)
+        {
+            _logger.LogInformation($"Requesting page {request.Page} for stars {request.Stars}");
```

Comment:
Our logging library supports formatting:

```suggestion
            _logger.LogInformation("Requesting page {Page} for stars <= {Stars}", request.Page, request.Stars);
```

String formatting is CPU heavy, this allows the logging library to avoid work if the log line won't be recorded. In addition, some logging libraries support "structured" logging and don't format the string in the app process.